### PR TITLE
Added Docker Support

### PR DESCRIPTION
**Pull request by dylhack**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+WORKDIR /app
+# NOTE(dylhack): this is an intentional possible exception to make sure
+#                modmail is configured before executed.
+COPY config.yml /app/config.yml
+COPY . /app
+
+RUN [ "npm", "i" ]
+RUN [ "npm", "run", "build" ]
+
+CMD [ "npm", "run", "start" ]
```
Comment (AltF02) [inline Dockerfile]: Create a runtime and build container:

```suggestion
FROM node:17-alpine AS base

WORKDIR /opt/app
COPY config.yml /app/config.yml
COPY . /opt/app

# Change this to yarn.lock for yarn
COPY --chown=node:node package-lock.json . 
COPY --chown=node:node package.json .
RUN npm install

FROM base AS build

COPY . /opt/app

RUN npm run build

FROM base AS prod

ENV NODE_ENV="production"

COPY --from=build /opt/app/dist /opt/app/dist
COPY --from=build /opt/app/node_modules /opt/app/node_modules
COPY --from=build /opt/app/package.json /opt/app/package.json

RUN chown node:node /opt/app

USER node

CMD [ "npm", "run", "start" ]
```

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+.env
+.github
+.gitignore
+.eslintrc.json
+.gitignore
+config.example.yml
+CONTRIBUTING.md
+FUNDING.yml
+package-lock.json
+yarn.lock
```
Comment (AltF02) [inline .dockerignore]: Remove these

**Review by AltF02 — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  server:
+    build: .
+    ports:
+      - "3001:80"
+    links:
+      - postgres 
+  postgres:
+    image: postgres:14.2-alpine
+    environment:
+      - POSTGRES_PASSWORD=postgres
```
Comment (AltF02) [inline docker-compose.yml]: Add a volume for postgres for easy backup

**Review by AltF02 — CHANGES_REQUESTED** [review]

**Comment by AltF02** [comment]
Consider adding an Kubernetes Resources file for easier deployment if we go k8s or k3s. Doubt we will tho so its optional

File: config.docker.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+database:
+  # The database config should work out 
+  # of the box without modifying these values.
+  user: 'postgres'
+  password: 'postgres'
+  host: 'postgres'
+  port: 5432
+  database: 'postgres'
+  # Max connections
+  max: 20
```
Comment (AltF02) [inline config.docker.yml:28]: Consider adding an check to server to check if appkey has been changed from the default

**Comment by AltF02** [comment]
Merge master into this

File: .dockerignore (hunk 2 above)
Comment (dylhack) [reply .dockerignore]: resolved in 092eaf5fa97954086c3b8661a1f2dce9d119a68a

File: config.docker.yml (hunk 4 above)
Comment (dylhack) [reply config.docker.yml:28]: out of scope, won't fix.

File: docker-compose.yml (hunk 3 above)
Comment (dylhack) [reply docker-compose.yml]: resolved in fb81ae312181916a2c29d9d98c79dd38dbd9cf22 0d3ce742ebf059964b97372281cd14b01da7c762

File: src/database/database.ts [hunk 5]
```diff
@@ -49,7 +55,40 @@ export default class DatabaseManager {
     }
 
     public async init(): Promise<void> {
-      const client = await this.pool.connect();
+      const log = ModmailBot.getLogger('database-init');
+      let attempts = 0;
+      let client: PoolClient | null = null;
+      let finalError: Error | null = null;
```
Comment (AltF02) [inline src/database/database.ts]: Nit: add space

File: src/database/database.ts [hunk 6]
```diff
@@ -49,7 +55,40 @@ export default class DatabaseManager {
...
+      const reconnect = () => new Promise<PoolClient>((res) => {
+        setTimeout(async () => {
+          res(await this.pool.connect());
+        }, DatabaseManager.RECON_DELAY_MS);
+      });
+
+      do {
+        try {
```
Comment (AltF02) [inline src/database/database.ts]: Add debug logging

File: src/database/database.ts [hunk 7]
```diff
@@ -49,7 +55,40 @@ export default class DatabaseManager {
...
+            log.warn(
+              'Failed to connect to postgres, trying again in'
+              + ` ${DatabaseManager.RECON_DELAY_MS / 1000} seconds.`,
+            );
+            // eslint-disable-next-line no-await-in-loop
+            client = await reconnect();
+          }
+        } catch (err) {
+          const e = err as Error;
+          finalError = e;
```
Comment (AltF02) [inline src/database/database.ts]: Can be simplified to finalError = err as Error;

File: src/database/database.ts [hunk 8]
```diff
@@ -49,7 +55,40 @@ export default class DatabaseManager {
...
+        }
+        attempts += 1;
+      } while (attempts < DatabaseManager.ATTEMPTS && client === null);
```
Comment (AltF02) [inline src/database/database.ts]: Why not use a for loop?

**Review by AltF02 — CHANGES_REQUESTED** [review]

File: src/database/database.ts (hunk 8 above)
Comment (dylhack) [reply src/database/database.ts]: outdated as of eae2f1f9b7705e89454ab1288800dda9588d3491

File: src/database/database.ts (hunk 7 above)
Comment (dylhack) [reply src/database/database.ts]: outdated as of eae2f1f9b7705e89454ab1288800dda9588d3491

File: src/database/database.ts [hunk 9]
```diff
@@ -48,9 +54,30 @@ export default class DatabaseManager {
       this.permissions = new PermissionsTable(pool);
     }
 
-    public async init(): Promise<void> {
-      const client = await this.pool.connect();
+    public async init(delay = 0, attempts = 0): Promise<void> {
```
Comment (AltF02) [inline src/database/database.ts]: Change this to an higher default delay

**Review by AltF02 — CHANGES_REQUESTED** [review]

File: src/database/database.ts (hunk 6 above)
Comment (dylhack) [reply src/database/database.ts]: resolved by 3a901141ae7a373b44ac6daef7ac0d4d5c6d24f4

File: src/database/database.ts (hunk 5 above)
Comment (dylhack) [reply src/database/database.ts]: you're a space

File: src/database/database.ts (hunk 9 above)
Comment (dylhack) [reply src/database/database.ts]: resolved: 845e3ae2da7a65591bc398ad4a9f0a57e5220952

**Comment by AltF02** [comment]
To not halter development I'm merging this, I will modify the dockerfile later in a PR to improve

**Pull request merged**